### ISSUE TITLE
Expose _ddata or another form of low level, locale aware memory buffer

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#16797]
Our current strategy for standard library containers requires three essential pieces which we do not expose to users today:

- A low level move (handled by https://github.com/chapel-lang/chapel/issues/16172)
- A way to manually deinitialize elements (handled by https://github.com/chapel-lang/chapel/issues/16172)
- A low level memory block

Depending on how discussion in https://github.com/chapel-lang/chapel/issues/16250 plays out, it may be possible for users to comfortably use arrays as backing store for their own container implementations.

However, I still think it would be useful to expose a lightweight, locale aware buffer to users who wish to build collections without using heavyweight arrays and domains.

Such a buffer would:

- Not default initialize elements
- Not be aware of which elements are initialized and which are not
- Not deinitialize elements upon deinit
- Store its own size

Additionally, such a buffer _could_:

- Provide convenience methods to move-initialize / deinit an element
- Provide convenience methods to move-initialize / deinit a range of elements
- Provide optional bounds checking

We could store this buffer in a module called `LowLevelBuffer`, or store it in `LowLevelMove` (and perhaps rename that module to something like `LowLevelTools`).

```chapel

import LowLevelBuffer;

class C { var x = 0; }

// E.g. allocate a new low level buffer with size 8
const hi = 8;
var buf = new LowLevelBuffer.buffer(shared C, hi);

// Prints '8'
writeln(buf.size);

// NO WARNING, ticket to segfault city!
writeln(buf);

// Go ahead and initialize buffer elements...
for i in 0..<hi do buf.moveInitialize(i, new shared C(i));

// Now a writeln makes sense...
writeln(buf);

// Simulate a collection operation, deinit the last 4 elements...
buf.deinitRange(4..7);

// When 'buf' goes out of scope, the first 4 elements won't be deinitialized.
// We have to do it ourselves...
buf.deinitRange(0..3);
```

**Comment by mppf** [comment]
Just noting that issues #12490 and #8680 are related.

> Depending on how discussion in #16250 plays out, it may be possible for users to comfortably use arrays as backing store for their own container implementations.

I don't think that containers need to wait for an answer to #16250 in order to use an array instead of a `_ddata` or `c_ptr` to a buffer. They can `noinit` the array and use the functionality described in #16172 and possibly #16173.

I think there is some agreement that data structure authors should be able to use arrays for this if they wish to. Whether or not something like `_ddata` is also available to them is not something I think there is agreement upon. (My own opinion is that something like `_ddata` should be available.)

**Comment by e-kayrakli** [comment]
Can we use or repurpose https://chapel-lang.org/docs/master/modules/packages/Buffers.html for this?

**Comment by dlongnecke-cray** [comment]
If we decide that we don't care about the `Buffers` package and that we don't want to put a `buffer` into the `LowLevelMove` module, I think we totally could, yeah.

**Comment by mppf** [comment]
I think we should ignore the package module when coming up with names. If we want to name the standard module `Buffers`, we can rename (or delete) the package module.

**Comment by dlongnecke-cray** [comment]
Should we perhaps add support for a type like `c_wide_ptr` or a locale aware `c_ptr` (related: https://github.com/chapel-lang/chapel/issues/12490). In that thread it seemed like @benharsh was proposing that we expose `_ddata` as a sort of wide pointer instead of necessarily a (locale aware) pointer to a buffer.

This thread is advocating for some sort of low level buffer, e.g.:

```chapel

record buffer {
  type T;
  var data: _ddata(T);
  var size: int;
}
```

However it seems to me like something like `c_wide_ptr` would be more general. We could introduce implicit coercions from `c_ptr` to `c_wide_ptr` and just let users `c_malloc` and manage buffers themselves.

**Comment by dlongnecke-cray** [comment]
On a related note (@bradcray), working on solutions to the resized non-nilable array debacle got me thinking...

I am not sure if we have a public issue for this, but I remember you advocating for using arrays as low-level buffers in the past. Up until recently I treated that as one of my design constraints.

However, I now feel more strongly that something like exposing `_ddata` or providing a `c_wide_ptr` is preferable to having users build collections out of arrays.

Core operations on arrays like parallel iteration operate on the assumption that all elements of the array are initialized. When using an array as backing store some/many elements remain uninitialized. Nothing about the array type communicates this inconsistent state to the user. We could expend some effort to add a "buffer-mode" to the array type that turns things like parallel iteration off, but I think a cleaner solution would be to introduce another, much more simple type. It might be a wrapper around `_ddata`, or a construct like `c_wide_ptr`, or maybe we just encourage use of `c_ptr`.

My guess is that this is at least part of your goal with #17202? (If we can use `c_ptr` for arrays, then we can have more confidence that users can use it for their own collections.)

**Comment by mppf** [comment]
Re parallel iteration on an array with uninitialized elements - I would expect it is fine to do this; it's just that you shouldn't read the values of uninitialized elements when doing so. In that way I think it's a more general problem. So I'd put it this way:

When some elements are uninitialized, the code working with it needs to behave differently, and so it's nice to have a different type reflecting that.

**Comment by mppf** [comment]
I think that we need a user-facing equivalent do `_ddata`. This comment had a bunch of suggested names - https://github.com/chapel-lang/chapel/issues/12490#issuecomment-480505878 . Coming up with the name for it seems to me to be the main challenge here.

I do not know if we will want to rename `c_ptr` to, say, `cPtr` (because we normally use camel case).

Anyway, I'm fairly happy with `c_wide_ptr` as a low-level name for `_ddata`, FWIW. I know it's not really for interoperability so the `c_` part is sortof a misnomer, but it is a low-level ("C level" ?) thing so it still makes sense to me.

**Comment by dlongnecke-cray** [comment]
I have some confusion about `c_wide_ptr` vs `_ddata`.

I thought that `_ddata` was a wide pointer to a dynamically allocated buffer of some size. And even if re-aliased to point to a different `_ddata`, that's still going to be a wide pointer to a dynamically allocated buffer.

Wouldn't we want the concept of `c_wide_ptr` to be more general? In that it's just a wide pointer - we could have it alias anything? So it might be pointing to a dynamic buffer, a stack allocated element, etc...

**Comment by dlongnecke-cray** [comment]
I think I might be conflating the functions `_ddata_allocate` and `_ddata_free` with the `_ddata` type itself. It just so happens that those functions create a `_ddata` that points to a dynamically allocated buffer, but there's nothing keeping us from adding functionality to point to other stuff if we wanted to.

Confirm or deny?

**Comment by mppf** [comment]
> I think I might be conflating the functions `_ddata_allocate` and `_ddata_free` with the `_ddata` type itself. It just so happens that those functions create a `_ddata` that points to a dynamically allocated buffer, but there's nothing keeping us from adding functionality to point to other stuff if we wanted to.

That's my understanding.

**Comment by bradcray** [comment]
> I remember you advocating for using arrays as low-level buffers in the past. Up until recently I treated that as one of my design constraints.

My concern that led me to say this is mostly the following:  that if we end up building a language in which data structures can't be built in terms of other data structures, but have to always go back to a primitive C-like buffer in order to define themselves, to me, that feels like a red flag.  E.g., if a user were to say "I want to build a stack type from an array or a list" or "I want to build a new sparse array which I think of as being composed of three 1D arrays" or "I want to build a tiled array implementation that uses a series of 2D arrays"  and our answer was always "Sorry, you need to build those from 1D C buffers instead," it suggests to me that something is wrong.

I definitely agree with creating a better user-facing replacement for _ddata (and using it ourselves where we use _ddata).  I'm not crazy about calling it `c_wide_ptr` because (as Michael notes and I've mentioned before) I think of `c_*` types as being about interoperability with C and I don't view C as having wide pointers.  So I'd probably go with a name like `widePtr` or `chplPtr` or something instead.  That also like the notion of a `buffer` type like David shows, which stores a pointer + size.  I think this relates a bit to requests for a `c_dynamic_array` type similar to the current `c_array` type (issue #16169).  And I think it's fine if users create data structures from these types; I just don't think that should be a necessity.

**Comment by dlongnecke-cray** [comment]
That suggests to me that maybe we should consider more strongly some sort of "buffer mode" for the array type in which you can't do something like `forall`.

**Comment by mppf** [comment]
> in which you can't do something like forall

I don't see what doing `forall` over the array has to do with it. Is the issue just that some of the memory might be uninitialized? If so, that problem applies to `for` as well as `forall`; you might use `forall` to initialize some/all elements; and it also applies to element access like `writeln(A[i])`.

**Comment by mppf** [comment]
A sortof related issue is that the AtomicObject module needs to be able to create a wide pointer of a particular type. If we had something like `c_wide_ptr` it would make sense to be able to cast it to an `unmanaged C?` (or to cast it to `c_void_ptr` first and then cast that). Today, our ability to cast a C pointer to a Chapel object is limited to single-locale, unless you do some hacky things.

**Comment by dlongnecke-cray** [comment]
That's true, the problem does apply to `for` as well as `forall`. I fixate on `forall` because a user might inclined to use it for their own implementation rather than write their own (it seems the harder thing to write when compared to a `for`).

I consider indexing a more fundamental property than a loop (if you don't have it, you can't do anything with the buffer).

If we were to expose `_ddata`, as a type named `buffer`, the interface that I would propose would only offer indexing. The implication behind such a simple type is that users are required to write all the other methods themselves.

**Comment by bradcray** [comment]
>  fixate on forall because a user might inclined to use it for their own implementation rather than write their own (it seems the harder thing to write when compared to a for).

I don't know that I agree with that (and was going to say so, but Michael beat me to it).  If I'm writing a queue data type using an array or buffer or whatever beneath it, even taking initialized vs. not initialized out of the equation, it wouldn't be correct behavior for my forall to iterate over all of the allocated array elements if only 2/3 of them are occupied by the queue (I'm imagining I'm not keeping the array the exact size that's needed by the queue, but using recursive doubling or something like that such that array.size > queue.size).

**Comment by dlongnecke-cray** [comment]
Right. All I'm trying to say that it's easier to deal with users accidentally using the wrong tool for the job if it's just not possible to do that thing in the first place. 

I suppose as long as the whole program can blow up in your face if you `A[i]` an uninitialized element, we might as well not care overly much about a `for` or `forall` blowing up as well.
